libqrexec/exec.c

```diff
@@ -105,4 +147,232 @@ void do_fork_exec(const char *cmdline, int *pid, int *stdin_fd, int *stdout_fd,
+        abort();
+    }
+    const struct qrexec_parsed_command command = {
+       .username = username,
+       .command = realcmd,
+       .service_descriptor = service_descriptor,
+       .service_descriptor_length = service_descriptor_length,
+    };
+    return execute_parsed_qubes_rpc_command(&command, pid, stdin_fd, stdout_fd, stderr_fd, stdin_buffer);
+}
+
+static int execute_parsed_qubes_rpc_command(
+        const struct qrexec_parsed_command *const command, int *const pid, int *const stdin_fd,
+        int *const stdout_fd, int *const stderr_fd, struct buffer *stdin_buffer) {
+    char const *const delimiter = memchr(command->service_descriptor, '+', command->service_descriptor_length);
+    size_t const service_length = delimiter ?
+        (size_t)(delimiter - command->service_descriptor) : command->service_descriptor_length;
+
+    if (!service_length) {
+        fputs("Service path empty\n", stderr);
+        return -1;
+    } else if (service_length > NAME_MAX) {
+        fprintf(stderr, "Service path too long to execute: %zu\n", service_length);
+        return -1;
+    }
+
+    int s;
+    if ((s = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
+        perror("socket");
+        return -1;
+    }
+
+    bool use_bare_path = command->service_descriptor_length > NAME_MAX;
+    for (;;) {
+        for (size_t i = 0; i < ARRAY_SIZE(qubes_rpc_directories); ++i) {
+            static_assert(sizeof "/usr/local/etc/qubes-rpc/" + NAME_MAX < QUBES_SOCKADDR_UN_MAX_PATH_LEN,
+                          "buffer length too low");
+            char service_full_path[QUBES_SOCKADDR_UN_MAX_PATH_LEN] = { 0 };
+            size_t const directory_length = qubes_rpc_directories[i].length;
+
+            // The total size of the path (not including NUL terminator).
+            // Cannot overflow due to earlier checks.
+            size_t const total_path_length = directory_length +
+                (use_bare_path ? service_length : command->service_descriptor_length);
+            if (sizeof service_full_path <= total_path_length) {
+                fputs("qrexec internal error: miscalculated the buffer size needed", stderr);
+                abort();
+            }
+            memcpy(service_full_path, qubes_rpc_directories[i].string, directory_length);
+            memcpy(service_full_path + directory_length, command->service_descriptor, total_path_length - directory_length);
+            service_full_path[total_path_length] = '\0';
+
+            if (!qubes_connect(s, service_full_path, total_path_length)) {
+                *stdout_fd = *stdin_fd = s;
+                if (stderr_fd) {
+                    *stderr_fd = -1;
+                }
+                *pid = -1;
+                set_nonblock(s);
+                buffer_append(stdin_buffer, command->service_descriptor, strlen(command->service_descriptor) + 1);
+                return 0;
+            }
+            switch (errno) {
```

Comment:
I think calling `connect()` first makes this much harder than it needs to be. This is a pretty complicated bit of control flow: infinite loop (`for(;;)`) with changing the `use_bare_path` parameter, and possible exits through `return`, `break`, and `goto out`, depending on many possible error codes.

I'm finding it pretty hard to understand all the possible outcomes, and still I think there are some weird corner cases: for instance, if a file is a socket but the listening process died (so we get `ECONNREFUSED`), we will fork and try execute it.

Why not first check if a file exists, and then decide what to do with it? Something like:

```c
// Look for `service_name+arg` and `service_name` files in both dirs, true
if (!find_service_file(command->service_descriptor, file_path)) {
    fprintf(stderr, "file not found for %s\n", command->service_descriptor);
    return 1;
}

if (is_a_socket(file_path)) {
    if (!qubes_connect(file_path) {
        perror("connect");
        return 1;
    }
    // prepare socket
    return 0;
} else if (is_executable(file_path)) {
     return do_fork_exec(file_path, ...);
} else {
     fprintf(stderr, "unrecognized service type (not a socket or executable): %s\n", file_path);
     return 1;
}
```

That would simplify the control flow (no loops over the whole code, just possibly a loop over directories in `find_service_file`), make sure that the errors (e.g. for `connect()`) are always reported, and match what `qubes-rpc-multiplexer` does (find a file, then execute it).

Reply:
(As discussed with @marmarek, I'm happy to leave this part until after merge; number 1 priority for me is getting it under unit tests).

Reply:
I actually agree with you.  I am not sure if this should go live until this is fixed.